### PR TITLE
chore: allow tokens to be provided via secure cookie instead

### PR DESCRIPTION
**Pull request by mackrorysd**
## Description

Collaboration between myself and @billboggs. When using the "external sessions" configuration in managed clusters, the JWT was previously passed via the URL. There was a technical reason this was necessary initially, but this is no longer the case and this is undesirable because of the potential for users to paste URLs or screenshots that include their token. Using cookies will be more secure, so instead the token is now placed in a cookie (this does require the cookie to be set with an appropriate domain for the cluster - and it can be set as a "secure" cookie with HttpOnly set to true).

I'm open to changing the name of the cookie if we need to, but as some browser have tight limits on cookie sizes, I think `jwt` is an appropriate balance of clarity and brevity. 

## Test Plan

I manually tested both the CLI and web UI against:
- Regular Determined / MLDE cluster run via `devcluster`
- Cluster configured with external sessions

No changes are required for the CLI, so this change is backwards-compatible with old releases - it will set the Authorization header but that still works and there are no security concerns with that.

## Checklist

- [X] Changes have been manually QA'd
- [N/A] User-facing API changes need the "User-facing API Change" label.
- [N/A] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [N/A] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

File: master/internal/grpcutil/auth.go [hunk 1]
```diff
@@ -157,7 +158,7 @@ func GetUser(ctx context.Context) (*model.User, *model.UserSession, error) {
 			return nil, nil, ErrPermissionDenied
 		}
 		return userModel, session, nil
-	case sql.ErrNoRows, db.ErrNotFound:
+	case sql.ErrNoRows, db.ErrNotFound, jwt.ErrTokenExpired:
```
Comment (stoksc) [inline master/internal/grpcutil/auth.go:161]: for my understanding, why did we need this? looking at `UserByToken`, i would think the `v2.Verify` call implodes already and you get a `db.ErrNotFound` for expires.

Comment (stoksc) [reply master/internal/grpcutil/auth.go:161]: yes, i understand that. when i say 'db error' i mean this ~~dumb~~ wonderful place where we swallow the verification error and return a db error https://github.com/determined-ai/determined/blob/36dac4ba5c3516c978c246702659446b886e97f0/master/internal/user/postgres_users.go#L126

Comment (stoksc) [reply master/internal/grpcutil/auth.go:161]: but i see now this probably branches before that due to the external token with saas.

Comment (stoksc) [reply master/internal/grpcutil/auth.go:161]: but overall, i was just trying to understand why this change needs this extra case, or it is a bug fix/only tangentially related.

File: master/internal/grpcutil/api.go [hunk 2]
```diff
@@ -115,7 +115,9 @@ func RegisterHTTPProxy(ctx context.Context, e *echo.Echo, port int, cert *tls.Ce
 	handler := func(c echo.Context) error {
 		request := c.Request()
 		if c.Request().Header.Get("Authorization") == "" {
-			if cookie, err := c.Cookie(cookieName); err == nil {
+			if cookie, err := c.Cookie("jwt"); err == nil {
```
[the remark this answers is not in the document]
Comment (stoksc) [reply master/internal/grpcutil/api.go:118]: uhh.. not sure, tbh, those are good points. certainly cant name it `hpe_mlde_or_mlds_v1_or_v2_jwt` then..

[the remark this answers is not in the document]
Comment (stoksc) [reply master/internal/grpcutil/api.go:118]: even just `djwt` to disambiguate it kinda makes sense to me, but i dont have strong opinions

**Review by stoksc — APPROVED** [review]
lgtm, meant to approve before, have nothing blocking, just qs

**Review by stoksc — APPROVED** [review]

File: master/internal/grpcutil/api.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ioga) [reply master/internal/grpcutil/api.go:118]: `det_jwt`?

File: master/internal/grpcutil/auth.go (hunk 1 above)
Comment (mackrorysd) [reply master/internal/grpcutil/auth.go:161]: Oh this is actually in connection with a related bug in MLDE code, which is being discussed by @billboggs. Part of reason is that the `jwt` package has some unusual handling of it's own error types - bit of a trade-off between making sure we return the correct error code so not everything looks like a server error and authentication problems can be handled appropriately, and wanting to keep implementation details of UserByExternalToken from leaking out.

**Comment by mackrorysd** [comment]
Changes:
- JWT is renamed to `det_jwt`
- Error handling tweak - no changes required to MLDE now
- If det_jwt is present, it'll override anything else. Possibly a temporary change if we want it to be just because there seems to be ways to `undefined` to end up being set as the Authorization header from the front-end. We've found and fixed some causes for this, but it still happens once in a while. It's correct behavior IMO to say that det_jwt should take precedence whenever it is set.

**Pull request merged**